<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jenna Landy">
<meta name="dcterms.date" content="2024-02-09">

<title>Jenna Landy - Getting Started with Mutational Signatures</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../images/Logo.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Jenna Landy</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../research.html" rel="" target="">
 <span class="menu-text">Research</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../software.html" rel="" target="">
 <span class="menu-text">Software</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../resources.html" rel="" target="">
 <span class="menu-text">Resources</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#biological-motivation" id="toc-biological-motivation" class="nav-link active" data-scroll-target="#biological-motivation">Biological Motivation</a></li>
  <li><a href="#defining-mutational-signatures" id="toc-defining-mutational-signatures" class="nav-link" data-scroll-target="#defining-mutational-signatures">Defining Mutational Signatures</a>
  <ul class="collapse">
  <li><a href="#data-mutational-catalog-m" id="toc-data-mutational-catalog-m" class="nav-link" data-scroll-target="#data-mutational-catalog-m">Data: Mutational Catalog <span class="math inline">\(M\)</span></a></li>
  <li><a href="#latent-factors-mutational-signatures-p" id="toc-latent-factors-mutational-signatures-p" class="nav-link" data-scroll-target="#latent-factors-mutational-signatures-p">Latent Factors: Mutational Signatures <span class="math inline">\(P\)</span></a></li>
  </ul></li>
  <li><a href="#statistical-methods" id="toc-statistical-methods" class="nav-link" data-scroll-target="#statistical-methods">Statistical Methods</a></li>
  <li><a href="#interpretation" id="toc-interpretation" class="nav-link" data-scroll-target="#interpretation">Interpretation</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Getting Started with Mutational Signatures</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jenna Landy </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 9, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="logo.png" class="img-fluid figure-img" width="300"></p>
</figure>
</div>
<p>Mutational signatures analysis models mutational processes in cancer with latent factors. This method characterizes cancer genomes in a way that is <strong><em>interpretable</em></strong> and <strong><em>clinically relevant</em></strong>. This article serves as an introduction to the biological motivation for mutational signatures and the statistical methods used.</p>
<section id="biological-motivation" class="level2">
<h2 class="anchored" data-anchor-id="biological-motivation">Biological Motivation</h2>
<p>Until recently, cancer diagnosis and treatment was based primarily on the tissue of origin, clinical presentation, and tumor morphology. The ability to sequence tumor DNA has changed this. The term “cancer” now refers to many diseases “differentiated on the basis of varying combinations of cancer gene mutations” <span class="citation" data-cites="baez-ortega2017">(<a href="#ref-baez-ortega2017" role="doc-biblioref">Baez-Ortega and Gori 2017</a>)</span>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="over_time.png" class="img-fluid figure-img" width="300"></p>
</figure>
</div>
<p>Cancer is characterized by uncontrolled cell replication due to mutations in somatic, or non-germline, cells <span class="citation" data-cites="pfeifer2010">(<a href="#ref-pfeifer2010" role="doc-biblioref">Pfeifer 2010</a>)</span>. There are some well-studied biological processes that lead to somatic mutations. For instance, in non-melanoma skin cancers, sunlight exposure leaves its trace through C-to-T transitions at dipyrimidine sequences and CC-to-TT tandem mutations. On the other hand, in smoking-associated lung cancers, we see G-to-T transversions, particularly at methylated CpG sites in TP53 <span class="citation" data-cites="pfeifer2010">(<a href="#ref-pfeifer2010" role="doc-biblioref">Pfeifer 2010</a>)</span>. Other processes are not due to external exposures, including DNA damage repair deficiencies. However, we are still learning many processes that cause somatic mutations in cancer <span class="citation" data-cites="alexandrov2013">(<a href="#ref-alexandrov2013" role="doc-biblioref">Alexandrov et al. 2013</a>)</span>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mut_process.png" class="img-fluid figure-img" width="300"></p>
</figure>
</div>
<p>Though they hadn’t yet been defined, Stratton, Cambelle, &amp; Futreal perfectly summarized the biological motivation for mutational signatures in 2009 <span class="citation" data-cites="stratton2009">(<a href="#ref-stratton2009" role="doc-biblioref">Stratton, Campbell, and Futreal 2009</a>)</span>:</p>
<blockquote class="blockquote">
<p>the catalogue of somatic mutations present in a cancer cell&nbsp;… represents a cumulative archaeological record of all the mutational processes the cancer cell has experienced throughout the lifetime of the patient. It provides a rich, and predominantly unmined, source of information for cancer epidemiologists and biologists with which to interrogate the development of individual tumors</p>
</blockquote>
<p>Mutational signatures analysis models a tumor’s mutational landscape as the composition of multiple mutational processes working at once.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mut_landscape.png" class="img-fluid figure-img" width="300"></p>
</figure>
</div>
</section>
<section id="defining-mutational-signatures" class="level2">
<h2 class="anchored" data-anchor-id="defining-mutational-signatures">Defining Mutational Signatures</h2>
<section id="data-mutational-catalog-m" class="level3">
<h3 class="anchored" data-anchor-id="data-mutational-catalog-m">Data: Mutational Catalog <span class="math inline">\(M\)</span></h3>
<p>Most signatures work focus on single base substitution (SBS) mutations, though signatures have been developed for double base substitutions, small insertions and deletions, copy number variations, structural variation, and RNA single base substitutions <span class="citation" data-cites="tate2018">(<a href="#ref-tate2018" role="doc-biblioref">Tate et al. 2018</a>)</span>.</p>
<p>For single base substitutions (SBS), a “type” of mutation is defined by the mutation and the nucleotides on either side. For instance, ACT &gt; AGT has a C&gt;T substitution (the mutation) with an A on the left and T on the right (the context). In the simplest case, we’re not concerned with which DNA strand the mutation was on, so we only consider substitutions at an A or a C. This gives six options for the substitution A&gt;C, A&gt;G, A&gt;T, C&gt;A, C&gt;G, C&gt;T, four options for the nucleotide on the left, and four options for the nucleotide on the right, giving 6 ⋅ 4 ⋅ 4 = <strong><em>96 SBS mutation types</em></strong>. We use base-pair matching to convert mutations at a T or a G into one of these 96 mutation types.</p>
<p>A tumor’s <strong><em>mutational catalog</em></strong> is its count of each mutation type. The catalog for each tumor genome <span class="math inline">\(g\)</span>, <span class="math inline">\(M_{g}\)</span>, is a vector of length 96. The full mutational catalog matrix <span class="math inline">\(M\)</span> has dimension 96 x G where G is the total number of tumor genomes, and tumors’ catalogs are columns.</p>
<div style="text-align: center; font-weight: bold;">
<p><span class="math inline">\(M\)</span> = Mutational catalog matrix, our observed data</p>
<p><span class="math inline">\(M_{kg}\)</span> = counts of mutation type <span class="math inline">\(k\)</span> present in tumor genome <span class="math inline">\(g\)</span></p>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="M_matrix.png" class="img-fluid figure-img" width="300"></p>
</figure>
</div>
</section>
<section id="latent-factors-mutational-signatures-p" class="level3">
<h3 class="anchored" data-anchor-id="latent-factors-mutational-signatures-p">Latent Factors: Mutational Signatures <span class="math inline">\(P\)</span></h3>
<p>A <strong>mutational signature</strong> is a mathematical representation of a mutational process. A signature defines <strong>how likely the mutational process is to give rise to each type of mutation</strong>. Mathematically, signature <strong>n</strong>, denoted <span class="math inline">\(\mathbf{P_n}\)</span>, is a probability distribution over these mutation types. The full signatures matrix <strong>P</strong> has dimension 96 x N where N is the total number of signatures, and signatures are columns.</p>
<div style="text-align: center; font-weight: bold;">
<p><span class="math inline">\(P\)</span> = Signature matrix, our latent factors representing mutational processes</p>
<p><span class="math inline">\(P_{kn}\)</span> = probability of mutation type <span class="math inline">\(k\)</span> resulting from mutational signature <span class="math inline">\(n\)</span></p>
<p><span class="math inline">\(P_{kn}\)</span> = Pr(mutation type k | signature n) such that <span class="math inline">\(\sum_k P_{kn} = 1 \forall n\)</span></p>
</div>
<p>A common way to visualize individual signatures is as a bar plot of these probabilities, grouped based on the central single base substitution, each with 16 options for context. Below is this type of plot for one signature, <a href="https://cancer.sanger.ac.uk/signatures/sbs/sbs3/">COSMIC SBS3</a>.</p>
<div style="text-align: center;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="SBS3.jpg" class="img-fluid figure-img" width="500"></p>
<figcaption class="figure-caption">COSMIC SBS3 GRCh37 (signatures v3.4) <span class="citation" data-cites="tate2018">(<a href="#ref-tate2018" role="doc-biblioref">Tate et al. 2018</a>)</span></figcaption>
</figure>
</div>
</div>
</section>
</section>
<section id="statistical-methods" class="level2">
<h2 class="anchored" data-anchor-id="statistical-methods">Statistical Methods</h2>
<p>In order to model mutational signatures, we need to convert our biological motivation about mutational processes into a mathematical expression using our mathematical definition of mutational signatures.</p>
<p>Biologically, we said</p>
<blockquote class="blockquote">
<p><em>a tumor’s mutational landscape is the composition of multiple mutational processes working at once</em>.</p>
</blockquote>
<p>Mathematically, we can say</p>
<blockquote class="blockquote">
<p><em>a tumor’s mutational catalog is a linear combination of multiple mutational signatures</em>.</p>
</blockquote>
<p>The weights of this linear combination are defined by a matrix <span class="math inline">\(E\)</span>, where <span class="math inline">\({E}_{ng}\)</span> is the <strong>number of mutations</strong> in tumor genome g attributed to mutational signature n.&nbsp;<span class="math inline">\(E_{ng}\)</span> is often called the <strong>exposure</strong> of tumor genome g to mutational signature n or equivalently the <strong>contribution</strong> of mutational signature n to tumor genome g. This definition simplifies to the expression</p>
<p><span class="math display">\[M = PE.\]</span></p>
<p>Because <span class="math inline">\(M\)</span> is a counts matrix, and <span class="math inline">\(P\)</span> a matrix of probabilities, both are inherently non-negative. We define <span class="math inline">\(E\)</span> as non-negative as well, because mutational processes can only add mutations, not get rid of them<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. Although we originally defined signatures <span class="math inline">\(P_n\)</span> as probability distributions summing to one, we can rescale matrices after the fact rather than incorporating the additional constraint here.</p>
<p><span class="math display">\[
\begin{aligned}
M_g =&amp; \sum_n P_n E_{ng}\\
M = PE, &amp;\text{ all non-negative}
\end{aligned}
\]</span></p>
<p>We’ve turned the biological idea of mutational signatures into a simple matrix decomposition <span class="math inline">\(M = PE\)</span> with all non-negative matrices. This can be solved with <strong>non-negative matrix factorization</strong> (NMF).&nbsp;</p>
</section>
<section id="interpretation" class="level2">




</section>


<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">Interpretation</h2><div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-alexandrov2013" class="csl-entry" role="listitem">
Alexandrov, Ludmil B., Serena Nik-Zainal, David C. Wedge, Samuel A. J. R. Aparicio, Sam Behjati, Andrew V. Biankin, Graham R. Bignell, et al. 2013. <span>“Signatures of Mutational Processes in Human Cancer.”</span> <em>Nature</em> 500 (7463): 415–21. <a href="https://doi.org/10.1038/nature12477">https://doi.org/10.1038/nature12477</a>.
</div>
<div id="ref-baez-ortega2017" class="csl-entry" role="listitem">
Baez-Ortega, Adrian, and Kevin Gori. 2017. <span>“Computational Approaches for Discovery of Mutational Signatures in Cancer.”</span> <em>Briefings in Bioinformatics</em> 20 (1): 77–88. <a href="https://doi.org/10.1093/bib/bbx082">https://doi.org/10.1093/bib/bbx082</a>.
</div>
<div id="ref-pfeifer2010" class="csl-entry" role="listitem">
Pfeifer, Gerd P. 2010. <span>“Environmental Exposures and Mutational Patterns of Cancer Genomes.”</span> <em>Genome Medicine</em> 2 (8): 54. <a href="https://doi.org/10.1186/gm175">https://doi.org/10.1186/gm175</a>.
</div>
<div id="ref-stratton2009" class="csl-entry" role="listitem">
Stratton, Michael R., Peter J. Campbell, and P. Andrew Futreal. 2009. <span>“The Cancer Genome.”</span> <em>Nature</em> 458 (7239): 719–24. <a href="https://doi.org/10.1038/nature07943">https://doi.org/10.1038/nature07943</a>.
</div>
<div id="ref-tate2018" class="csl-entry" role="listitem">
Tate, John G, Sally Bamford, Harry C Jubb, Zbyslaw Sondka, David M Beare, Nidhi Bindal, Harry Boutselakis, et al. 2018. <span>“COSMIC: The Catalogue Of Somatic Mutations In Cancer.”</span> <em>Nucleic Acids Research</em> 47 (D1): D941–47. <a href="https://doi.org/10.1093/nar/gky1015">https://doi.org/10.1093/nar/gky1015</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>The use positive weights for mutational signatures is different from other matrix decomposition or dictionary learning applications to genomics. For example, if we were looking at RNA-seq data, a mutational process may increase <em>or</em> decrease gene expression, so negative weights would be allowed.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>