<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Jenna Landy - Getting Started with Mutational Signatures</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../images/Logo.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Jenna Landy</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../research.html" rel="" target="">
 <span class="menu-text">Research</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../software.html" rel="" target="">
 <span class="menu-text">Software</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../resources.html" rel="" target="">
 <span class="menu-text">Resources</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Getting Started with Mutational Signatures</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>Mutational signatures analysis is a new way to characterize cancer genomes in a way that is interpretable and clinically relevant. Most methods are based on non-negative matrix factorization (NMF). This article serves as an introduction to the biological motivation for mutational signatures, the statistical methods used, and a sample analysis on simulated data.</p>
<p><strong>Table of&nbsp;Contents</strong></p>
<ol type="1">
<li><p>Biological Motivation</p></li>
<li><p>Defining Mutational Signatures</p></li>
<li><p>Statistical Methods</p></li>
<li><p>Sample Analysis</p></li>
</ol>
<hr>
<section id="biological-motivation" class="level3">
<h3 class="anchored" data-anchor-id="biological-motivation">Biological Motivation</h3>
<p>Until recently, cancer diagnosis and treatment was based primarily on the tissue of origin, clinical presentation, and tumor morphology. The ability to sequence tumor DNA has changed this. The term “cancer” now refers to many diseases “differentiated on the basis of varying combinations of cancer gene mutations” [<a href="https://doi.org/10.1186/s12885-019-5677-2">Van Hoeck et al.&nbsp;2019</a>].</p>
<p>Cancer is characterized by uncontrolled cell replication due to mutations in somatic, or non-germline, cells [<a href="https://doi.org/10.1093/bib/bbx082">Baez-Ortega and Gori 2019</a>]. There are some well-studied biological processes that lead to somatic mutations. For instance, in non-melanoma skin cancers, sunlight exposure leaves its trace through C-to-T transitions at dipyrimidine sequences and CC-to-TT tandem mutations. On the other hand, in smoking-associated lung cancers, we see G-to-T transversions, particularly at methylated CpG sites in TP53 (linked to polycyclic aromatic hydrocarbons) [<a href="https://doi.org/10.1186/gm175">Pfeifer 2010</a>]. However, we are still in the dark as far as <em>most</em> of the processes that cause somatic mutations in cancer [<a href="https://doi.org/10.1038/nature12477">Alexandrov et al.&nbsp;2013</a>].</p>
<p>Though they hadn’t yet been defined, <a href="https://doi.org/10.1038/nature07943">Stratton, Cambelle, &amp; Futreal (2009)</a> perfectly summarizes the biological motivation for mutational signatures:</p>
<blockquote class="blockquote">
<p>the catalogue of somatic mutations present in a cancer cell&nbsp;… represents a cumulative archaeological record of all the mutational processes the cancer cell has experienced throughout the lifetime of the patient. It provides a rich, and predominantly unmined, source of information for cancer epidemiologists and biologists with which to interrogate the development of individual tumors</p>
</blockquote>
<p><strong>Mutational signatures analysis is a recent development to model a tumor’s mutational landscape as the composition of multiple mutational processes working at once.</strong></p>
<p>Most signatures focus on single base substitution (SBS) mutations, though signatures have been developed for double base substitutions, small insertions and deletions, copy number variations, structural variation, and RNA single base substitutions [<a href="https://doi.org/10.1093/nar/gky1015">Tate et al.&nbsp;2019</a>].</p>
<hr>
</section>
<section id="defining-mutational-signatures" class="level3">
<h3 class="anchored" data-anchor-id="defining-mutational-signatures">Defining Mutational Signatures</h3>
<ul>
<li><p><strong>M =</strong> Mutational catalog<br>
<strong>M[k,g]</strong> = counts of SBS mutation type <strong>k</strong> present in tumor genome <strong>g</strong></p></li>
<li><p><strong>P =</strong> Signature matrix<br>
<strong>P[k,n]</strong> = probability of SBS mutation type <strong>k</strong> resulting from mutational signature <strong>n</strong></p></li>
</ul>
<p>For SBS signatures, a “type” of mutation is defined by the single base substitution and the nucleotides on either side. For instance, ACT &gt; AGT has a C&gt;T substitution (the mutation) and an A on the left and T on the right (the context<strong>)</strong>. In the simplest case, we’re not concerned with what DNA strand the mutation was on, so we only consider substitutions of an A or a C. This gives six options for the substitution A&gt;C, A&gt;G, A&gt;T, C&gt;A, C&gt;G, C&gt;T, four options for the nucleotide on the left, and four options for the nucleotide on the right, giving 6 ⋅ 4 ⋅ 4 = <strong>96 SBS mutation types</strong>.</p>
<p>A tumor’s <strong>mutational catalog</strong> is its count of each mutation type. The catalog for each tumor genome <strong>g</strong>, <strong>M[&nbsp;,g]</strong>, is a vector of length 96. The full mutational catalog matrix <strong>M</strong> has dimension 96 x G where G is the total number of tumor genomes, and tumors’ catalogs are columns.</p>
<p>A <strong>mutational signature</strong> is a mathematical representation of a mutational process. A signature defines <strong>how likely the mutational process is to give rise to each type of mutation</strong>. Mathematically, signature <strong>n</strong>, denoted <strong>P[,n]</strong>, is a probability distribution over these mutation types. The full signatures matrix <strong>P</strong> has dimension 96 x N where N is the total number of signatures, and signatures are columns.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://cdn-images-1.medium.com/max/1600/1*9u3TKHJWeT08lmodTYPmDQ.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Definition of a mutational signature</figcaption>
</figure>
</div>
<p>Below is an example of a common visualization of a signature (<a href="https://cancer.sanger.ac.uk/signatures/sbs/sbs3/">COSMIC SBS3</a> here). Mutation types are grouped based on the central single base substitution, each with 16 options for context.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://cdn-images-1.medium.com/max/1600/1*NrD-g3vtgFriPUhzq_JS8g.jpeg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption"><a href="https://cancer.sanger.ac.uk/signatures/sbs/sbs3/">COSMIC SBS3 GRCh37 (signatures v3.4)</a> [<a href="https://doi.org/10.1093/nar/gky1015">Tate et al.&nbsp;2019</a>]</figcaption>
</figure>
</div>
<hr>
</section>
<section id="statistical-methods" class="level3">
<h3 class="anchored" data-anchor-id="statistical-methods">Statistical Methods</h3>
<p>Biologically, we said that <em>a tumor’s mutational landscape is the composition of multiple mutational processes working at once</em>. Mathematically, we represent this by defining <em>a tumor’s mutational catalog as a linear combination of multiple mutational signatures</em>. We define each weight <strong>E[n,g]</strong> as the <strong>contribution of</strong> mutational signature <strong>n</strong> to tumor genome <strong>g</strong>, or equivalently the <strong>exposure</strong> of tumor genome <strong>g</strong> to mutational signature <strong>n</strong>. This definition simplifies to the expression <strong>M = PE</strong>.</p>
<p>Because <strong>M</strong> is a counts matrix, and <strong>P</strong> a matrix of probabilities, both are inherently non-negative. We define <strong>E</strong> as non-negative as well, because mutational processes can only add mutations, not get rid of them*. Although we originally defined signatures <strong>P[,n]</strong> as probability distributions summing to one, we can rescale matrices after the fact rather than incorporating the addition constraint here.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://cdn-images-1.medium.com/max/1600/1*4t0_S5xs650OvgSUhX5GDw.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Non-negative matrix factorization (NMF) for mutational signatures</figcaption>
</figure>
</div>
<p>The matrix decomposition <strong>M = PE</strong> with all non-negative matrices can be solved with <strong>non-negative matrix factorization</strong> (NMF).&nbsp;</p>
<p>* I note that the use positive weights is different from other matrix decomposition or dictionary learning applications to genomics. For example, if we were looking at RNA-seq data, a mutational process may increase <em>or</em> decrease gene expression, so weights would be allowed to be negative.</p>
</section>
<section id="simulated-data-sample-analysis" class="level3">
<h3 class="anchored" data-anchor-id="simulated-data-sample-analysis">Simulated Data &amp; Sample&nbsp;Analysis</h3>
<section id="generating-simulated-data" class="level4">
<h4 class="anchored" data-anchor-id="generating-simulated-data">Generating Simulated Data</h4>
<p>We can generate simulated data based off of the current “gold standard” signatures from COSMIC [<a href="https://doi.org/10.1093/nar/gky1015">Tate et al.&nbsp;2019</a>]. Below, I use their first five signatures and generate 20 tumor genomes.</p>
<pre><code>N &lt;- 5
G &lt;- 20

# download COSMIC signatures, use the first N as our P
url &lt;- "https://cog.sanger.ac.uk/cosmic-signatures-production/documents/COSMIC_v3.3.1_SBS_GRCh37.txt"
cosmic_signatures &lt;- read.csv(url, sep = '\t', row.names = 1)
P &lt;- as.matrix(cosmic_signatures[,1:N])

# simulate # mutations from each signature n to each tumor genome g
# centered s.t. each tumor has around 1000 total mutations
E &lt;- matrix(nrow = N, ncol = G)
n_mutations &lt;- 1000
for (g in 1:G) {
    E[,g] &lt;- rpois(N, 1000/N)
}

# simulate mutational catalog centered at PE
M = matrix(nrow = 96, ncol = G)
for (g in 1:G) {
    M[,g] &lt;- rpois(96, P %*% E[,g])
}</code></pre>
</section>
<section id="analysis-using-nmf" class="level4">
<h4 class="anchored" data-anchor-id="analysis-using-nmf">Analysis Using&nbsp;NMF</h4>
<p>Traditionally, optimal <strong>P</strong> and <strong>E</strong> matrices are found through gradient descent using multiplicative rather than additive updates to ensure non-negativity. In R, gradient descent NMF can be performed using the NMF package on CRAN [<a href="https://doi.org/10.1186/1471-2105-11-367">Gaujoux and Seoighe 2010</a>].&nbsp;</p>
<pre><code>install.packages("NMF")
library(NMF)</code></pre>
<p>More recently, NMF has been translated to the Bayesian space, where priors are put on <strong>P</strong> and <strong>E</strong> and an optimal solution is found by maximizing the posteriors of <strong>P</strong> and <strong>E</strong> using MCMC methods (I will be publishing articles on the specifics of these methods soon!). In R, Bayesian NMF can be fit using my bayesNMF package on GitHub [<a href="https://github.com/jennalandy/bayesNMF">Landy 2023</a>].</p>
<pre><code>library(devtools)
devtools::install_github("jennalandy/bayesNMF")
library(bayesNMF)</code></pre>
<p>For&nbsp;either method, we must specify the reduced dimension rank, or in our case, the number of mutational signatures to fit. Although we know the true number of signatures is 5 because the data is simulated, I will proceed as if this is unknown to reflect a real data analysis.</p>
<p>We can pass in ranks from 2 up to 15 to the NMF package in order to choose a rank. This is always a rather subjective choice, and there are several approaches to choose the optimal rank. Many introductory tutorials will reference the elbow method, or choosing the rank that is at the inflection point of the RSS curve. A more sophisticated approach focuses on Brunet’s cophenetic correlation cofficient (CCC) method. The CCC measures dispersion for the consensus matrix. This method chooses the rank after which the CCC begins to decrease (in our data, at rank = 5, top left plot below).</p>
<pre><code>nmf_res &lt;- NMF::nmf(adeno_catalog, rank = 2:15)
plot(nmf_res)</code></pre>
<hr>
<p>Moving forward with rank = 5, we can run both gradient descent and Bayesian methods:</p>
<pre><code>gd_nmf_rank7_res &lt;- NMF::nmf(M, rank = 5)
gd_nmf_rank7_P &lt;- nmf_rank7_res@fit@W
gd_nmf_rank7_E &lt;- nmf_rank7_res@fit@H

# This function saves results in bayes_nmf_rank7.RData and 
# logs progress in bayes_nmf_rank7.log
bayes_nmf_rank7_res &lt;- bayesNMF::nmf_normal_exponential(
  M, N = 5,
  file = "bayes_nmf_rank7"
)
bayes_nmf_rank7_P &lt;- bayes_nmf_rank7_res$P.mean
bayes_nmf_rank7_E &lt;- bayes_nmf_rank7_res$E.mean</code></pre>
</section>
<section id="evaluate-results" class="level4">
<h4 class="anchored" data-anchor-id="evaluate-results">Evaluate Results</h4>
<p>Finally, we can investigate how our estimated mutational signatures align with the true signatures used for simulating the data.</p>
<pre><code>bayesNMF::get_heatmap(est_P = gd_nmf_rank7_P, true_P = P)</code></pre>
<pre><code>bayesNMF::get_heatmap(est_P = bayes_nmf_rank7_P, true_P = P)</code></pre>
<hr>
</section>
</section>
<section id="references" class="level3">
<h3 class="anchored" data-anchor-id="references">References</h3>
<ol type="1">
<li><p>Van Hoeck, A., Tjoonk, N.H., van Boxtel, R. <em>et al.</em> Portrait of a cancer: mutational signature analyses for cancer diagnostics. <em>BMC Cancer</em> <strong>19</strong>, 457 (2019). <a href="https://doi.org/10.1186/s12885-019-5677-2" class="uri">https://doi.org/10.1186/s12885-019-5677-2</a>.</p></li>
<li><p>Helleday, Thomas, Saeed Eshtad, and Serena Nik-Zainal. 2014. “Mechanisms Underlying Mutational Signatures in Human Cancers.” <em>Nature Reviews Genetics</em> 15: 585–98. <a href="https://doi.org/10.1038/nrg3729" class="uri">https://doi.org/10.1038/nrg3729</a>.</p></li>
<li><p>Adrian Baez-Ortega, Kevin Gori, Computational approaches for discovery of mutational signatures in cancer, <em>Briefings in Bioinformatics</em>, Volume 20, Issue 1, January 2019, Pages 77–88, <a href="https://doi.org/10.1093/bib/bbx082" class="uri">https://doi.org/10.1093/bib/bbx082</a></p></li>
<li><p>Alexandrov, Ludmil B., Serena Nik-Zainal, David C. Wedge, Samuel A. J. R. Aparicio, Sam Behjati, Andrew V. Biankin, Graham R. Bignell, et al.&nbsp;2013. “Signatures of Mutational Processes in Human Cancer.” <em>Nature</em> 500: 415–21. <a href="https://doi.org/10.1038/nature12477" class="uri">https://doi.org/10.1038/nature12477</a>.</p></li>
<li><p>Pfeifer, G.P. Environmental exposures and mutational patterns of cancer genomes. <em>Genome Med</em> <strong>2</strong>, 54 (2010). <a href="https://doi.org/10.1186/gm175" class="uri">https://doi.org/10.1186/gm175</a></p></li>
<li><p>Stratton, M., Campbell, P. &amp; Futreal, P. The cancer genome. <em>Nature</em> <strong>458</strong>, 719–724 (2009). <a href="https://doi.org/10.1038/nature07943" class="uri">https://doi.org/10.1038/nature07943</a></p></li>
<li><p>John G Tate, Sally Bamford, Harry C Jubb, Zbyslaw Sondka, David M Beare, Nidhi Bindal, Harry Boutselakis, Charlotte G Cole, Celestino Creatore, Elisabeth Dawson, Peter Fish, Bhavana Harsha, Charlie Hathaway, Steve C Jupe, Chai Yin Kok, Kate Noble, Laura Ponting, Christopher C Ramshaw, Claire E Rye, Helen E Speedy, Ray Stefancsik, Sam L Thompson, Shicai Wang, Sari Ward, Peter J Campbell, Simon A Forbes, COSMIC: the Catalogue Of Somatic Mutations In Cancer, <em>Nucleic Acids Research</em>, Volume 47, Issue D1, 08 January 2019, Pages D941–D947, <a href="https://doi.org/10.1093/nar/gky1015" class="uri">https://doi.org/10.1093/nar/gky1015</a></p></li>
<li><p>Gaujoux R, Seoighe C (2010). “A flexible R package for nonnegative matrix factorization.” <em>BMC Bioinformatics</em>, <strong>11</strong>(1), 367. ISSN 1471–2105, <a href="https://doi.org/10.1186/1471-2105-11-367">doi:10.1186/1471–2105–11–367</a>, <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-367" class="uri">https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-367</a>.</p></li>
<li><p>Landy, Jenna (2023). “bayesNMF: an R package with various models for Bayesian NMF.” <em>GitHub</em>. <a href="https://github.com/jennalandy/bayesNMF" class="uri">https://github.com/jennalandy/bayesNMF</a></p></li>
</ol>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>